### PR TITLE
Support -reporter=sarif

### PR DESCRIPTION
**Pull request by haya14busa**
- [x] Updated Unreleased section in [CHANGELOG](https://github.com/reviewdog/reviewdog/blob/master/CHANGELOG.md) or it's not notable changes.

### CODE REVIEW

File: comment_iowriter.go [hunk 1]
```diff
@@ -142,3 +144,135 @@ func (cw *RDJSONCommentWriter) Flush(_ context.Context) error {
 	}
 	return nil
 }
+
+var _ CommentService = &SARIFCommentWriter{}
+
+// SARIFCommentWriter
```
Comment (review-dog) [inline comment_iowriter.go:150]: **[golint-circleci]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>comment on exported type SARIFCommentWriter should be of the form "SARIFCommentWriter ..." (with optional leading article)

File: comment_iowriter.go [hunk 2]
```diff
@@ -142,3 +144,135 @@ func (cw *RDJSONCommentWriter) Flush(_ context.Context) error {
...
+type SARIFCommentWriter struct {
+	w        io.Writer
+	comments []*Comment
+	toolName string
+}
+
+func NewSARIFCommentWriter(w io.Writer, toolName string) *SARIFCommentWriter {
```
Comment (review-dog) [inline comment_iowriter.go:157]: **[golint-circleci]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>exported function NewSARIFCommentWriter should have comment or be unexported

File: comment_iowriter.go [hunk 3]
```diff
@@ -142,3 +144,135 @@ func (cw *RDJSONCommentWriter) Flush(_ context.Context) error {
...
+	return &SARIFCommentWriter{w: w, toolName: toolName}
+}
+
+func (cw *SARIFCommentWriter) Post(_ context.Context, c *Comment) error {
```
Comment (review-dog) [inline comment_iowriter.go:161]: **[golint-circleci]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>exported method SARIFCommentWriter.Post should have comment or be unexported

File: comment_iowriter.go [hunk 4]
```diff
@@ -142,3 +144,135 @@ func (cw *RDJSONCommentWriter) Flush(_ context.Context) error {
...
+	cw.comments = append(cw.comments, c)
+	return nil
+}
+
+func (cw *SARIFCommentWriter) Flush(_ context.Context) error {
```
Comment (review-dog) [inline comment_iowriter.go:166]: **[golint-circleci]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>exported method SARIFCommentWriter.Flush should have comment or be unexported

File: comment_iowriter.go [hunk 5]
```diff
@@ -142,3 +144,135 @@
...
```
[the remark this answers is not in the document]
Comment (haya14busa) [reply comment_iowriter.go:166]: Sample: We can discuss this Code scanning alert.

File: comment_iowriter.go [hunk 6]
```diff
@@ -142,3 +144,135 @@ func (cw *RDJSONCommentWriter) Flush(_ context.Context) error {
...
+			Message: sarif.Message{
+				Text: sarif.String(c.Result.Diagnostic.Message),
+			},
+		}
+		if code := c.Result.Diagnostic.GetCode(); code.GetValue() != "" {
+			result.RuleID = sarif.String(code.GetValue())
+			rules[code.GetValue()] = sarif.ReportingDescriptor{
+				ID:      code.GetValue(),
+				HelpURI: sarif.String(code.GetUrl()),
+			}
```
Comment (massongit) [inline comment_iowriter.go:186]:
```suggestion
			run.Tool.Driver.Rules = append(run.Tool.Driver.Rules, sarif.ReportingDescriptor{
				ID:      code.GetValue(),
				HelpURI: sarif.String(code.GetUrl()),
			})
```

Values of the variable `rules` are only used when assigning to `run.Tool.Driver.Rules`.
Therefore, I think it is better to assign to `run.Tool.Driver.Rules` here.

Comment (haya14busa) [reply comment_iowriter.go:186]: There can be multiple results with the same Rule ID. 
We should add one rule per rule id.

Comment (haya14busa) [reply comment_iowriter.go]: On second thought, yes, we can refactore the code a bit to make it clear. 
I refactore the code. 907b2c0d5e4fcc98aed49d09ca3426b306e1dd18 
Thanks for the suggestion!

**Review by massongit — APPROVED** [review]
LGTM

**Pull request merged**